### PR TITLE
Supporting glob-style operations

### PR DESCRIPTION
**Pull request by choldgraf**
OK this one has bugged me for long enough that I decided to try and actually implement a fix.

This PR adds support for `glob`-module-style pattern matching, so you should be able to do `--ignore **/*ignore*.txt` or `--ignore **/ignore` and it should just work.

I added a test that I _think_ checks the edge-cases here, but welcome to have thoughts on how to improve.

- closes https://github.com/executablebooks/sphinx-autobuild/issues/112
- closes https://github.com/executablebooks/sphinx-autobuild/issues/117
- might be a fix for https://github.com/executablebooks/sphinx-autobuild/issues/120

maybe @asmeurer or @consideRatio would be willing to give a look? I think both of you have gotten bitten by this one before as well :-)

Note that I also ran `pre-commit autoupdate` and ran it on all our files, because I think an outdated version was causing pre-commit to fail.

### CODE REVIEW

File: noxfile.py [hunk 1]
```diff
@@ -44,6 +54,13 @@ def docs(session):
...
-        "sphinx-autobuild", "-b", "html", "docs/", "build/docs", *session.posargs
-    )
+    cmd = [
+        "sphinx-autobuild",
+        "-b",
+        "html",
+        "docs/",
+        "build/docs",
+        "--port",
+        "0",
```
Comment (consideRatio) [inline noxfile.py:64]: Just a refactoring for readability. To put flags and the associated value on the same line is a practice I've come to appreciate.

```suggestion
        "-b=html",
        "--port=0",
        "docs/",
        "build/docs",
```

**Review by consideRatio — APPROVED** [review]
This LGTM!

I have a gut feeling that the larger ignore logic where this is a part could be refactored and reduce in complexity, but I think that is out of scope for this PR and especially not something to address now if I can't propose something concrete.

File: noxfile.py (hunk 1 above)
Comment (choldgraf) [reply noxfile.py:64]: ah nice - I didn't realize that the `=` still worked in this case. Black was clobbering the "two arguments per line" setup I had before.

**Comment by asmeurer** [comment]
I'm not currently using autosummary, so it's not easy for me to test that use-case. 

However, I have had an issue with sphinx-autobuild ignores in general, which may or may not be related. In the SymPy docs, we have a lot of plots that are generated by the matplotlib directive. Sometimes, `sphinx-autobuild` gets stuck in a loop thinking that every single one of these files is updated, like

```
[sphinx-autobuild] Detected change: /Users/aaronmeurer/Documents/Python/sympy/sympy/doc/_build/plot_directive/modules/plotting-35.pdf
```

This happens even if I do `--ignore=_build/` (I tested it and it happens in this branch too).

If you want to reproduce this in the [SymPy](https://github.com/sympy/sympy) docs, you can install the dependencies from `doc/requirements.txt` then, from the `doc/` directory run

```
make livehtml # Runs sphinx-autobuild
```

then in another terminal

```
make clean
```

(the docs take a couple of minutes to build from scratch)

The infinite loop honestly wouldn't be so bad, except the webserver gets blocked by it, meaning if you try to click a link in the live webpage, it tries to wait for the build to finish before it loads the page, and thus never loads. Making the webserver not block like this in general would be helpful. 

Happy to open a separate issue for this if you'd like.

**Comment by asmeurer** [comment]
This sort of looping also happens whenever I checkout a new branch, one at a time for every file in the source repo that was changed. So I guess there are three problems here:

- sphinx-autobuild is not ignoring _build/, even when told to do so with --ignore
- The webserver blocks while a build is happening. This is particularly annoying when this looping happens, but it's also annoying in general. Could it be made asynchronous?
- When a bunch of files change, it does a bunch of rebuilds, one for each file. It ought to be able to detect changes in a batch to avoid unnecessary duplicate rebuilds.

**Comment by choldgraf** [comment]
Hmm - I'm not sure why your build folder isn't being ignored ..indeed I would have expected this PR to fix it. I can try to take another look and see if i may have missed something.

For the async and batch blocking stuff unfortunately i have no idea how that stuff works, and also have no real knowledge of this codebase other than what I learned while making this PR, so i think we should just open an issue to track it over time. (We also need to find a way to make a release, this repository got moved into executable books without transferring the ability to upload to pypi)

**Comment by choldgraf** [comment]
Another quick thought: can you double check that the path to your `_build` folder is correct when you pass the `--ignore` argument? I didn't see `--ignore` in [your makefile](https://github.com/sympy/sympy/blob/15295494bc150f631c093b89537fa9b29fd3fe81/doc/Makefile#L131-L132). What was the specific call that you use when using `--ignore`?

I opened two issues to track two of your requests above:

- #124 
- #125 

For the "`_build`" folder issue, I couldn't replicate it locally in a toy site I put together. I made a little Sphinx site with this structure:

- `docs`
- `build`
- And a `noxfile.py` to run `sphinx-autobuild`

I then set `conf.py` in `docs` to generate a folder with 10 files in it in `build/tmp` at build time.

I then used this invocation of sphinx-autobuild:

```
sphinx-autobuild -b html docs/ build/html --ignore build --port 0
```

and it correctly ignored the build folder. So I'm not sure why you aren't getting that behavior.

**Comment by asmeurer** [comment]
Yes, I never actually committed `--ignore` to the sympy Makefile because it never seemed to actually work. If you're testing this, add `--ignore=_build/` (and anyway, shouldn't the build folder be ignored automatically?). Here's the full line I am using to test this in the SymPy Makefile

```
sphinx-autobuild --open-browser --ignore=_build/ --watch .. --host $(LIVEHOST) --port $(LIVEPORT) -b html $(ALLSPHINXOPTS) $(SOURCEDIR) $(BUILDDIR)/html
```

**Comment by choldgraf** [comment]
I spent a while trying to get the sympy docs to build this morning but could never figure it out. I got a string of error messages that I had to resolve by installing certain system libraries etc, but the last one I ran into I didn't know how to debug `AttributeError: module 'unicodedata2' has no attribute 'ucnhash_CAPI'` and I ran out of time to work on this.

I am not sure why this doesn't work for you, but I think this PR at least solves some of the edge-cases that have been mentioned. I can remove the "closes" for your glob issue in case that is not resolved, but I'd suggest that we merge this one as an iterative improvement so that we can get it in before I go on paternity leave in a few days.

**Comment by asmeurer** [comment]
I've seen that error before. I'll have to see if I can remember what causes it.

**Comment by asmeurer** [comment]
And anyway, feel free to just open an issue for my problem. It's already a problem for me and this PR didn't seem to make it worse. Although I admit I didn't actually test the actual globbing nature of this PR.

**Comment by guideloom** [comment]
What are the chances this getting approved? This is much needed! :)

**Comment by choldgraf** [comment]
Sorry - I had a baby and dropped off the face of the earth. I can try to get back to this at some point but am on paternity leave for the time being

**Comment by asmeurer** [comment]
By the way, the repeated rebuilding issue doesn't just happen on SymPy. It seems to happen on any repo that uses autodoc with lots of Python files. Here's a simpler repo where I've also seen the issue that you can test https://github.com/quansight-labs/ndindex.

**Comment by jhgoebbert** [comment]
@choldgraf Thank you for the PR. It would be fantastic if it could be merged.

**Comment by choldgraf** [comment]
Hey all - I don't have time to work on this. I recommend somebody else picks up this PR. Feel free to fork and open a new one, or just take over this one

**Comment by jhgoebbert** [comment]
Thank you for the answer.  
I already picked it up and created a new PR which is rebased here https://github.com/sphinx-doc/sphinx-autobuild/pull/171
You can close this PR.

**Comment by choldgraf** [comment]
Done! For anybody else that comes across this PR, please check out its successor here:

- https://github.com/sphinx-doc/sphinx-autobuild/pull/171

Thanks @jhgoebbert !
